controllers/machinedeployment_controller.go

```diff
@@ -220,6 +220,15 @@ func (r *MachineDeploymentReconciler) reconcile(ctx context.Context, cluster *cl
 		return ctrl.Result{}, errors.Errorf("missing MachineDeployment strategy")
 	}
 
+	// TODO (alberto): store this in the reconciler.
+	now := metav1.Now()
+	if now.After(d.Spec.RolloutAfter.Time) &&
```

Comment:
Intentionally starting another discussion here, because it's a different idea.

So what this implementation does is it basically triggers a rollout because the template in the MD changes. I wonder if we could instead just include the check for rolloutafter at wherever we currently check if the MD has already bin rolled out (i.e. where we diff between the MD and the MS).

I'm not sure if there's a good way to do that. I suspect it would be more or less in `EqualMachineTemplate` (although the func name would have to be changed)

Reply:
I thought about that as well, but I couldn't come up with proper solution. `EqualMachineTemplate` is kind of a safeguard against hash collisions for the new/old MachineSet names. We could do:

```
if spec.rolloutAfter.before(now) then:
  include `spec.rolloutAfter.string()` as part of the MachineSet name hash
```

That way every time spec.rolloutAfter changes and is before Now the MachineSet name differs and so the MachineDeployment orchestrate the rolling update.

The issue with the above is that `mdutil.FindNewMachineSet(d, msList)` seems to prefer oldest MachineSet in case of equal templates.

Reply:
Okay thx. Let's continue in https://github.com/kubernetes-sigs/cluster-api/pull/4596#discussion_r705499279